torch/distributed/fsdp/_fully_shard/_fsdp_collectives.py

```diff
@@ -162,15 +165,16 @@ def foreach_all_gather(
             all_gather_inputs = [*chain.from_iterable(param_all_gather_inputs)]
         inp_split_sizes = [t.numel() for t in all_gather_inputs]
         all_gather_input_numel = sum(inp_split_sizes)
-        all_gather_input, all_gather_output = torch.ops.fsdp.all_gather_copy_in(
-            all_gather_inputs,
-            inp_split_sizes,
-            all_gather_input_numel,
-            world_size,
-            rank,
-            dtype,
-            device,
-        )
+        with torch.cuda.use_mem_pool(mempool) if mempool else nullcontext():
+            all_gather_input, all_gather_output = torch.ops.fsdp.all_gather_copy_in(
```

Comment:
I would have preferred to only apply the context manager to the low-level `torch.empty` calls, but in this case they are happening inside a `custom_op` and thus I wrap the whole op invocation.

It is in principle possible to pass the MemPool as an argument to the custom_op. It must be converted to a C++ custom class (see https://github.com/pytorch/pytorch/pull/151100). But, even then, some issues remain, notably in terms of dependencies between PyTorch libraries:
- the MemPool is defined by the CUDA caching allocator, which lives in c10 / cuda
- the C++ custom class tooling is defined in torch / cpu (with some helpers being in ATen / cpu)

I still haven't figured out a way to make this work in all cases. I can insist if people don't like the current compromise.